cloud/scope/machine.go

```diff
@@ -40,65 +40,61 @@ type MachineScopeParams struct {
 	AzureClients
 	Client       client.Client
 	Logger       logr.Logger
-	Cluster      *clusterv1.Cluster
 	Machine      *clusterv1.Machine
-	AzureCluster *infrav1.AzureCluster
 	AzureMachine *infrav1.AzureMachine
 }
 
-// NewMachineScope creates a new MachineScope from the supplied parameters.
-// This is meant to be called for each reconcile iteration.
-func NewMachineScope(params MachineScopeParams) (*MachineScope, error) {
-	if params.Client == nil {
-		return nil, errors.New("client is required when creating a MachineScope")
-	}
-	if params.Machine == nil {
-		return nil, errors.New("machine is required when creating a MachineScope")
-	}
-	if params.Cluster == nil {
-		return nil, errors.New("cluster is required when creating a MachineScope")
-	}
-	if params.AzureCluster == nil {
-		return nil, errors.New("azure cluster is required when creating a MachineScope")
-	}
-	if params.AzureMachine == nil {
-		return nil, errors.New("azure machine is required when creating a MachineScope")
-	}
-
-	if params.Logger == nil {
-		params.Logger = klogr.New()
-	}
-
-	helper, err := patch.NewHelper(params.AzureMachine, params.Client)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to init patch helper")
-	}
-	return &MachineScope{
-		client:       params.Client,
-		Cluster:      params.Cluster,
-		Machine:      params.Machine,
-		AzureCluster: params.AzureCluster,
-		AzureMachine: params.AzureMachine,
-		Logger:       params.Logger,
-		patchHelper:  helper,
-	}, nil
-}
-
 // MachineScope defines a scope defined around a machine and its cluster.
 type MachineScope struct {
 	logr.Logger
 	client      client.Client
 	patchHelper *patch.Helper
 
-	Cluster      *clusterv1.Cluster
+	ClusterScope azure.ClusterDescriber
 	Machine      *clusterv1.Machine
-	AzureCluster *infrav1.AzureCluster
 	AzureMachine *infrav1.AzureMachine
 }
 
-// Location returns the AzureMachine location.
+// PublicIPSpec returns the public IP specs.
+func (m *MachineScope) PublicIPSpecs() []azure.PublicIPSpec {
+	var spec []azure.PublicIPSpec
+	if m.AzureMachine.Spec.AllocatePublicIP == true {
+		nicName := azure.GenerateNICName(m.Name())
+		spec = append(spec, azure.PublicIPSpec{
+			Name: azure.GenerateNodePublicIPName(nicName),
+		})
+	}
+	return spec
+}
+
+// Location returns the AzureCluster location.
 func (m *MachineScope) Location() string {
-	return m.AzureCluster.Spec.Location
+	return m.ClusterScope.Location()
+}
+
+// ResourceGroup returns the AzureCluster resource group.
+func (m *MachineScope) ResourceGroup() string {
+	return m.ClusterScope.ResourceGroup()
```

Comment:
should we embed the struct/interface?

Reply:
Main reason is that this allows re-use of Services between AzureCluster, AzureMachine, etc. If we embed the ClusterScope then we have to do `scope.ResourceGroup()` for one and `scope.ClusterScope.ResourceGroup()` for the other. See publicips.go for an example of how that works (both AzureCluster and AzureMachine call publicips `Reconcile()`).

We also want to be able to have two different implementations for some of these shared functions but have the same signature, eg. `PublicIPSpecs()` returns a different spec for AzureCluster than for AzureMachine.